app/helpers/bulk_downloads_helper.rb

```diff
@@ -0,0 +1,27 @@
+module BulkDownloadsHelper
+  SAMPLE_NO_PERMISSION_ERROR = "You do not have permission to access all of the selected samples. Please contact us for help.".freeze
+  SAMPLE_STILL_RUNNING_ERROR = "Some selected samples have not finished running yet. Please remove these samples and try again.".freeze
+  SAMPLE_FAILED_ERROR = "Some selected samples failed their most recent run. Please remove these samples and try again.".freeze
+
+  def get_pipeline_run_ids_for_samples(sample_ids)
```

Comment:
Thanks for raising this. I didn't consider whether it should be in the pipeline run model.

After thinking about it, I do think this is related to bulk downloads. This gets valid pipeline runs, where "valid" is in the context of whether the pipeline run is ready to be part of a bulk download. I can rename the function and add a comment to make this more clear. The errors that are returned are also written for display in the bulk download modal.

You could argue that we should make this function more generic, into something like "get_successful_pipeline_run_ids_for_samples" in pipeline_runs model and then have this function call that one. But in bulk downloads we've chosen to fail if even one sample is invalid, and a generic function might want to just return all the valid pipeline run ids and not error out. We might relax the invalid constraint in bulk downloads in the future, but right now I think this function has enough unique requirements that it's fine to put in the bulk download helper.

Reply:
I think that should be a `strict mode` option which can raise or return the errors. There are a bunch of examples where this would be useful, and knowing about running or error samples is important even if treatment is different: heatmap, phylo tree creation, etc. I just fear that is yet another logic duplication.

Reply:
I also changed my comment from `related` -> `specific`.

Reply:
Great point about heatmap and other use cases. Okay, I'll make the change.